### PR TITLE
Fix build on OS X when GNU sed is present.

### PR DESCRIPTION
**Pull request by lorrin**
See #20147, which fixed it when system default BSD sed was present but did not account for possibility of GNU sed on OS X.

See also http://unix.stackexchange.com/questions/92895/how-to-achieve-portability-with-sed-i-in-place-editing

[Analytics]()

### CODE REVIEW

**Comment by lavalamp** [comment]
ok to test

@jbeda can you review, since you seem to be doing things in this area?

File: build/common.sh [hunk 1]
```diff
@@ -242,11 +242,15 @@ function kube::build::is_osx() {
   [[ "$(uname)" == "Darwin" ]]
 }
 
+function kube::build::is_gnu_sed() {
+  [[ $(sed --version 2>&1) == *GNU* ]]
```
Comment (jbeda) [inline build/common.sh]: Have you tested this on Mac OS X?  BSD based `sed` there gives an error and exits with 1 with `sed --version`.

```
$ sed --version
sed: illegal option -- -
usage: sed script [-Ealn] [-i extension] [file ...]
       sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]
$ echo $?
1
```

I'm afraid this test will trigger errexit on the mac.

Comment (lorrin) [inline build/common.sh]: Hi @jbeda , yes, I had tested on OS X stock `sed` and it worked correctly. Sorry for the slow reply; I had to do some digging to understand _why_ it worked. 😄 

`$()` command substitution passes on the `-1` exit code from OS X BSD sed. The answer, according to http://www.tldp.org/LDP/abs/html/options.html#OPTIONSREF, is that `set -e` will _Abort script at first error, when a command exits with non-zero status (**except in until or while loops, if-tests, list constructs**)_. So `[[ $(sed --version 2>&1) == *GNU* ]]` will return an error code but it is harmless, even when `set -e` is in play. Here's a minimal demonstration. Note that the script exits and the final echo is never printed.

```
bash-3.2$ cat test.sh
#!/bin/bash

[[ $(false) ]]
echo "Extended test false: $?"
$(false)
echo "Bare false: $?"

echo set -e
set -e

[[ $(false) ]]
echo "Extended test false: $?"
$(false)
echo "Bare false: $?"
```
```
bash-3.2$ ./test.sh
Extended test false: 1
Bare false: 1
set -e
Extended test false: 1
```

**Comment by jbeda** [comment]
Wow!  Very thorough.  Thanks for the analysis/explanation.

This LGTM.  

There are other places where we finesse the BSD/GNU issues (look at usages of tar) but I'm not sure that there is any common code to factor out.

Thanks!

**Pull request merged**